### PR TITLE
Extract common parts of ApplySnapshot

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: ref #8081

Problem Summary:

This is the second part of single big region snapshot handling. It mainly did the following:
1. extract `AsyncTasks` as a common util
2. gather `SSTFilesToBlockInputStream`'s arguments into `SSTFilesToBlockInputStreamOpts`
3. extract the three stream into `readFromStream`
4. make `readFromStream` return error code rather than exception for recoverable error

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/BoundedSSTFilesToBlockInputStream.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 PingCAP, Inc.
```
Comment (CalvinNeo) [inline dbms/src/Storages/DeltaMerge/BoundedSSTFilesToBlockInputStream.cpp:1]: Moved out from SSTFilesToBlockInputStream, with some header files excluded.

File: dbms/src/Storages/KVStore/Utils/AsyncTasks.h [hunk 2]
```diff
@@ -15,19 +15,15 @@
...
 
 #include <Common/UniThreadPool.h>
-#include <Storages/KVStore/FFI/ProxyFFI.h>
-#include <Storages/KVStore/MultiRaft/Disagg/FastAddPeer.h>
 
 #include <future>
 
 namespace DB
 {
+template <typename Key, typename Func, typename R>
```
Comment (CalvinNeo) [inline dbms/src/Storages/KVStore/Utils/AsyncTasks.h:23]: Intentionally, a `packaged_task` can be created from `Func`, which returns an `R` that is convertible from `result_of(Func)`.

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 3]
```diff
@@ -232,14 +279,39 @@ PrehandleResult KVStore::preHandleSSTsToDTFiles(
...
+                    term,
+                    index);
+            }
+
+            (void)table_drop_lock; // the table should not be dropped during ingesting file
+            break;
+        }
+        catch (DB::Exception & e)
+        {
+            if (e.code() == ErrorCodes::TABLE_IS_DROPPED)
```
Comment (CalvinNeo) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:297]: The `TABLE_IS_DROPPED` exception here can still be thrown by some `AtomicGetStorageSchema`.

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 4]
```diff
@@ -42,6 +42,96 @@ extern const int TABLE_IS_DROPPED;
...
+        return std::make_pair(
+            std::move(res),
+            PrehandleResult{
+                .ingest_ids = stream->outputFiles(),
+                .stats = PrehandleResult::Stats{
+                    .raft_snapshot_bytes = sst_stream->getProcessKeys().total_bytes(),
+                    .dt_disk_bytes = stream->getTotalBytesOnDisk(),
+                    .dt_total_bytes = stream->getTotalCommittedBytes()}});
+    }
+    catch (DB::Exception & e)
```
Comment (CalvinNeo) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:112]: We release sst_stream here

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 5]
```diff
@@ -137,69 +226,27 @@ PrehandleResult KVStore::preHandleSSTsToDTFiles(
...
-                    term,
-                    index);
-                stream->cancel();
-            }
-            prehandle_result.ingest_ids = stream->outputFiles();
-            prehandle_result.stats = PrehandleResult::Stats{
-                .raft_snapshot_bytes = sst_stream->getProcessKeys().total_bytes(),
-                .dt_disk_bytes = stream->getTotalBytesOnDisk(),
-                .dt_total_bytes = stream->getTotalCommittedBytes()};
+                DM::SSTFilesToBlockInputStreamOpts(opt));
```
Comment (JaySon-Huang) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:243]:
```suggestion
                std::move(opt));
```
I think we don't need this copying

File: dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.cpp
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.cpp:1]: Yes, seems some more files can be removed too

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp (hunk 5 above)
Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:243]: Why, opts is later used by `readFromStream`

Comment (JaySon-Huang) [reply dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:243]: OK, I miss that line

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 6]
```diff
@@ -42,6 +42,96 @@ extern const int TABLE_IS_DROPPED;
...
+};
+
+static inline std::tuple<ReadFromStreamResult, PrehandleResult> readFromStream(
+    RegionPtr new_region,
+    std::shared_ptr<std::atomic_bool> prehandle_task,
+    DM::FileConvertJobType job_type,
+    std::shared_ptr<StorageDeltaMerge> storage,
+    std::shared_ptr<DM::SSTFilesToBlockInputStream> sst_stream,
+    const DM::SSTFilesToBlockInputStreamOpts & opts,
+    TMTContext & tmt)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:66]:
```suggestion
static inline std::tuple<ReadFromStreamResult, PrehandleResult> readFromStream(
    const RegionPtr & new_region,
    const std::shared_ptr<std::atomic_bool> & prehandle_task,
    DM::FileConvertJobType job_type,
    const std::shared_ptr<StorageDeltaMerge> & storage,
    const std::shared_ptr<DM::SSTFilesToBlockInputStream> & sst_stream,
    const DM::SSTFilesToBlockInputStreamOpts & opts,
    TMTContext & tmt)
```

Comment (JaySon-Huang) [reply dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:66]: And maybe `executeTransform` is a better name for this function?

**Review by JaySon-Huang — APPROVED** [review]
The rest LGTM

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp (hunk 6 above)
Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp]: good idea

**Review by hongyunyan — APPROVED** [review]

**Pull request merged**